File: src/Player/Player.cpp
#include "Player/Player.h"
#include "World/World.h"
#include "Render/Camera.h"
#include <GLFW/glfw3.h>
#include <glm/gtc/matrix_transform.hpp>
#include <algorithm>
#include <cmath>

namespace mc {

namespace {
constexpr float GRAVITY = 28.0f;
constexpr float JUMP_VELOCITY = 8.6f;
constexpr float WALK_SPEED = 4.6f;
constexpr float SPRINT_SPEED = 6.2f;
constexpr float FLY_SPEED = 12.0f;
constexpr float WATER_DRAG = 0.6f;
}

Player::Player() {
    BlockId defaults[HOTBAR_SIZE] = {
        BLOCK_GRASS, BLOCK_DIRT, BLOCK_STONE, BLOCK_COBBLE, BLOCK_PLANKS,
        BLOCK_LOG, BLOCK_GLASS, BLOCK_GLOWSTONE, BLOCK_BRICK
    };
    for (int i = 0; i < HOTBAR_SIZE; ++i) hotbar[i] = defaults[i];
}

void Player::scrollHotbar(int dir) {
    selectedSlot = (selectedSlot - dir) % HOTBAR_SIZE;
    if (selectedSlot < 0) selectedSlot += HOTBAR_SIZE;
}

void Player::toggleMode() {
    mode = (mode == GameMode::Survival) ? GameMode::Creative
         : (mode == GameMode::Creative) ? GameMode::Spectator
                                        : GameMode::Survival;
    if (mode == GameMode::Spectator) flying = true;
    if (mode == GameMode::Survival) flying = false;
}

bool Player::collides(World& world, const glm::vec3& pos) const {
    if (mode == GameMode::Spectator) return false;
    glm::vec3 h = halfExtents();
    glm::vec3 mn = pos - glm::vec3(h.x, 0.0f, h.z);
    glm::vec3 mx = pos + glm::vec3(h.x, h.y * 2.0f, h.z);
    int x0 = (int)std::floor(mn.x), x1 = (int)std::floor(mx.x - 1e-4f);
    int y0 = (int)std::floor(mn.y), y1 = (int)std::floor(mx.y - 1e-4f);
    int z0 = (int)std::floor(mn.z), z1 = (int)std::floor(mx.z - 1e-4f);
    for (int y = y0; y <= y1; ++y)
        for (int z = z0; z <= z1; ++z)
            for (int x = x0; x <= x1; ++x) {
                BlockId b = world.getBlock(x, y, z);
                if (Blocks::isSolid(b)) return true;
            }
    return false;
}

void Player::resolveCollisions(World& world, glm::vec3& pos, glm::vec3& vel, float dt) {
    // Move axis by axis so we can slide along walls.
    glm::vec3 delta = vel * dt;

    pos.x += delta.x;
    if (collides(world, pos)) { pos.x -= delta.x; vel.x = 0; }

    pos.z += delta.z;
    if (collides(world, pos)) { pos.z -= delta.z; vel.z = 0; }

    pos.y += delta.y;
    if (collides(world, pos)) {
        pos.y -= delta.y;
        if (vel.y < 0) onGround = true;
        vel.y = 0;
    } else {
        onGround = false;
    }
}

void Player::applyMovement(World& world, const Camera& camera, const InputState& in, float dt) {
    // Determine if standing in water.
    glm::vec3 eye = position + glm::vec3(0, eyeHeight(), 0);
    inWater = world.getBlock((int)std::floor(position.x),
                             (int)std::floor(position.y + 0.5f),
                             (int)std::floor(position.z)) == BLOCK_WATER;

    glm::vec3 fwd = camera.forwardXZ();
    glm::vec3 rightDir = glm::normalize(glm::cross(fwd, glm::vec3(0, 1, 0)));
    glm::vec3 wish(0);
    if (in.forward) wish += fwd;
    if (in.back)    wish -= fwd;
    if (in.right)   wish += rightDir;
    if (in.left)    wish -= rightDir;
    if (glm::length(wish) > 0.0001f) wish = glm::normalize(wish);

    float speed = in.sprint ? SPRINT_SPEED : WALK_SPEED;

    if (flying) {
        speed = FLY_SPEED * (in.sprint ? 1.8f : 1.0f);
        velocity.x = wish.x * speed;
        velocity.z = wish.z * speed;
        velocity.y = 0;
        if (in.jump)  velocity.y += speed;
        if (in.sneak) velocity.y -= speed;
        resolveCollisions(world, position, velocity, dt);
        return;
    }

    // Horizontal acceleration toward wish velocity.
    glm::vec3 targetH = wish * speed;
    float accel = onGround ? 18.0f : 6.0f;
    velocity.x += (targetH.x - velocity.x) * std::min(1.0f, accel * dt);
    velocity.z += (targetH.z - velocity.z) * std::min(1.0f, accel * dt);

    // Gravity / water.
    if (inWater) {
        velocity.y -= GRAVITY * 0.3f * dt;
        velocity *= (1.0f - WATER_DRAG * dt);
        if (in.jump) velocity.y = 4.0f; // swim up
        velocity.y = std::max(velocity.y, -4.0f);
    } else {
        velocity.y -= GRAVITY * dt;
        if (in.jump && onGround) { velocity.y = JUMP_VELOCITY; onGround = false; }
    }
    velocity.y = std::max(velocity.y, -60.0f);

    resolveCollisions(world, position, velocity, dt);
}

void Player::updateStats(World& world, float dt) {
    if (mode != GameMode::Survival) { health = maxHealth; return; }

    // Fall damage.
    if (onGround && !wasOnGround_) {
        float fall = fallStartY_ - position.y;
        if (fall > 3.0f && !inWater) health -= (fall - 3.0f) * 1.0f;
    }
    if (!onGround && wasOnGround_) fallStartY_ = position.y;
    if (!onGround && position.y > fallStartY_) fallStartY_ = position.y;
    wasOnGround_ = onGround;

    // Environmental damage.
    int fx = (int)std::floor(position.x);
    int fy = (int)std::floor(position.y + 0.5f);
    int fz = (int)std::floor(position.z);
    BlockId feet = world.getBlock(fx, fy, fz);
    if (feet == BLOCK_CACTUS) health -= 2.0f * dt;

    // Suffocation if head is inside a solid block.
    BlockId head = world.getBlock(fx, (int)std::floor(position.y + eyeHeight()), fz);
    if (Blocks::isOpaque(head)) health -= 1.0f * dt;

    // Slow natural regen while fed.
    if (hunger > 6.0f && health < maxHealth) {
        regenTimer_ += dt;
        if (regenTimer_ > 2.0f) { health += 1.0f; regenTimer_ = 0.0f; }
    }
    hunger = std::max(0.0f, hunger - dt * 0.02f);
    if (hunger <= 0.0f) health -= dt * 0.5f;

    health = std::clamp(health, 0.0f, maxHealth);
    if (health <= 0.0f) dead = true;
}

void Player::update(World& world, Camera& camera, const InputState& in, float dt) {
    camera.addMouseDelta(in.mouseDX, in.mouseDY, mouseSensitivity);

    // Double-tap jump toggles flight (Creative only).
    if (in.jumpPressed && mode == GameMode::Creative) {
        double now = glfwGetTime();
        if (now - lastJumpTime_ < 0.30) flying = !flying;
        lastJumpTime_ = now;
    }

    if (!dead) {
        applyMovement(world, camera, in, dt);
        updateStats(world, dt);
    }

    camera.position = position + glm::vec3(0, eyeHeight(), 0);
}

void Player::respawn(World& world) {
    health = maxHealth;
    hunger = 20.0f;
    dead = false;
    velocity = glm::vec3(0);
    int h = world.findSpawnHeight(0, 0);
    position = glm::vec3(0.5f, (float)h, 0.5f);
}

} // namespace mc

File: minicraft-java/src/main/java/com/minicraft/player/Player.java
package com.minicraft.player;

import com.minicraft.render.Camera;
import com.minicraft.world.Blocks;
import com.minicraft.world.Inventory;
import com.minicraft.world.ItemStack;
import com.minicraft.world.Items;
import com.minicraft.world.World;
import org.joml.Vector3f;
import org.lwjgl.glfw.GLFW;

/** Player physics, collision, stats, and game modes. */
public class Player {
    public enum Mode { SURVIVAL, CREATIVE, SPECTATOR }

    public final Vector3f position = new Vector3f(0, 80, 0);
    public final Vector3f velocity = new Vector3f();
    public boolean onGround, flying, inWater, inLava;
    public Mode mode = Mode.CREATIVE;

    public float health = 20, maxHealth = 20, hunger = 20;
    public boolean dead;
    public float mouseSensitivity = 0.12f;

    public final Inventory inventory = new Inventory();

    private static final float GRAVITY = 28f, JUMP = 8.6f, WALK = 4.6f, SPRINT = 6.2f, FLY = 12f;

    private float fallStartY;
    private boolean wasOnGround = true;
    private double lastJumpTime = -10;
    private float regenTimer;

    /** The block id the selected hotbar item would place, or AIR if not placeable. */
    public int selectedBlock() {
        ItemStack s = inventory.selectedStack();
        if (s.isEmpty()) return Blocks.AIR;
        return Items.isBlock(s.item) ? Items.blockId(s.item) : Blocks.AIR;
    }

    public void scrollHotbar(int dir) { inventory.scroll(dir); }

    public void toggleMode() {
        mode = switch (mode) {
            case SURVIVAL -> Mode.CREATIVE;
            case CREATIVE -> Mode.SPECTATOR;
            case SPECTATOR -> Mode.SURVIVAL;
        };
        if (mode == Mode.SPECTATOR) flying = true;
        if (mode == Mode.SURVIVAL) flying = false;
    }

    public float halfX() { return 0.3f; }
    public float halfZ() { return 0.3f; }
    public float height() { return 1.8f; }
    public float eyeHeight() { return 1.62f; }

    private boolean collides(World w, Vector3f p) {
        if (mode == Mode.SPECTATOR) return false;
        float hx = halfX(), hz = halfZ();
        int x0 = (int) Math.floor(p.x - hx), x1 = (int) Math.floor(p.x + hx - 1e-4f);
        int y0 = (int) Math.floor(p.y), y1 = (int) Math.floor(p.y + height() - 1e-4f);
        int z0 = (int) Math.floor(p.z - hz), z1 = (int) Math.floor(p.z + hz - 1e-4f);
        for (int y = y0; y <= y1; y++)
            for (int z = z0; z <= z1; z++)
                for (int x = x0; x <= x1; x++)
                    if (Blocks.isSolid(w.getBlock(x, y, z))) return true;
        return false;
    }

    private void resolve(World w, float dt) {
        float dx = velocity.x * dt, dy = velocity.y * dt, dz = velocity.z * dt;
        position.x += dx;
        if (collides(w, position)) { position.x -= dx; velocity.x = 0; }
        position.z += dz;
        if (collides(w, position)) { position.z -= dz; velocity.z = 0; }
        position.y += dy;
        if (collides(w, position)) {
            position.y -= dy;
            if (velocity.y < 0) onGround = true;
            velocity.y = 0;
        } else onGround = false;
    }

    public void update(World w, Camera cam, InputState in, float dt) {
        if (in.jumpPressed && mode == Mode.CREATIVE) {
            double now = GLFW.glfwGetTime();
            if (now - lastJumpTime < 0.30) flying = !flying;
            lastJumpTime = now;
        }
        if (!dead) {
            move(w, cam, in, dt);
            updateStats(w, dt);
        }
        cam.position.set(position.x, position.y + eyeHeight(), position.z);
    }

    private void move(World w, Camera cam, InputState in, float dt) {
        int fx = (int) Math.floor(position.x), fz = (int) Math.floor(position.z);
        int fy = (int) Math.floor(position.y + 0.5f);
        inWater = w.getBlock(fx, fy, fz) == Blocks.WATER;
        inLava = w.getBlock(fx, fy, fz) == Blocks.LAVA;

        Vector3f fwd = cam.forwardXZ();
        Vector3f rightDir = new Vector3f(fwd).cross(new Vector3f(0, 1, 0)).normalize();
        Vector3f wish = new Vector3f();
        if (in.forward) wish.add(fwd);
        if (in.back) wish.sub(fwd);
        if (in.right) wish.add(rightDir);
        if (in.left) wish.sub(rightDir);
        if (wish.lengthSquared() > 1e-6f) wish.normalize();

        float speed = in.sprint ? SPRINT : WALK;
        if (flying) {
            speed = FLY * (in.sprint ? 1.8f : 1f);
            velocity.set(wish.x * speed, 0, wish.z * speed);
            if (in.jump) velocity.y += speed;
            if (in.sneak) velocity.y -= speed;
            resolve(w, dt);
            return;
        }

        float fluid = (inWater || inLava) ? 0.5f : 1f;
        float targetX = wish.x * speed * fluid, targetZ = wish.z * speed * fluid;
        float accel = onGround ? 18f : 6f;
        velocity.x += (targetX - velocity.x) * Math.min(1f, accel * dt);
        velocity.z += (targetZ - velocity.z) * Math.min(1f, accel * dt);

        if (inWater || inLava) {
            velocity.y -= GRAVITY * 0.3f * dt;
            velocity.mul(1f - 0.6f * dt);
            if (in.jump) velocity.y = 4f;
            if (velocity.y < -4f) velocity.y = -4f;
        } else {
            velocity.y -= GRAVITY * dt;
            if (in.jump && onGround) { velocity.y = JUMP; onGround = false; }
        }
        if (velocity.y < -60f) velocity.y = -60f;
        resolve(w, dt);
    }

    private void updateStats(World w, float dt) {
        if (mode != Mode.SURVIVAL) { health = maxHealth; return; }
        if (onGround && !wasOnGround) {
            float fall = fallStartY - position.y;
            if (fall > 3f && !inWater) health -= (fall - 3f);
        }
        if (!onGround && wasOnGround) fallStartY = position.y;
        if (!onGround && position.y > fallStartY) fallStartY = position.y;
        wasOnGround = onGround;

        int fx = (int) Math.floor(position.x), fz = (int) Math.floor(position.z);
        int fy = (int) Math.floor(position.y + 0.5f);
        int feet = w.getBlock(fx, fy, fz);
        if (feet == Blocks.CACTUS) health -= 2f * dt;
        if (inLava) health -= 4f * dt;
        int magmaBelow = w.getBlock(fx, (int) Math.floor(position.y - 0.1f), fz);
        if (magmaBelow == Blocks.MAGMA && onGround) health -= 1f * dt;

        int head = w.getBlock(fx, (int) Math.floor(position.y + eyeHeight()), fz);
        if (Blocks.isOpaque(head)) health -= 1f * dt;

        if (hunger > 6 && health < maxHealth) {
            regenTimer += dt;
            if (regenTimer > 2f) { health += 1; regenTimer = 0; }
        }
        hunger = Math.max(0, hunger - dt * 0.02f);
        if (hunger <= 0) health -= dt * 0.5f;
        health = Math.max(0, Math.min(maxHealth, health));
        if (health <= 0) dead = true;
    }

    public void respawn(World w) {
        health = maxHealth; hunger = 20; dead = false; velocity.set(0);
        int h = w.spawnHeight(0, 0);
        position.set(0.5f, h, 0.5f);
    }
}
